### PR TITLE
Improve choose/2 and int/2 generators

### LINKED ISSUE #65: int(Min, Max) shrinks to 0 instead of Min
Noticed that int(Min, Max) domain shrinks to 0 instead of Min. Is that expected?

For example:
```
125>  triq_dom:sampleshrink(triq_dom:int(5, 10)).
8
[7]
[6]
[5]
[4]
[3]
[2]
[1]
[0]
ok
```

If this is a bug, it looks 0 should be replaced with Min in this fragment:

https://github.com/triqng/triq/blob/master/src/triq_dom.erl#L607-L609

### PR DESCRIPTION
**Pull request by nickva**
~Fix int(Min, Max) to respect Min and Max limits when shrinking~

~The new behavior is try to shrink towards 0, but without going over the Min or~
~Max limits.~

(See discussion below. PR was updated with new behavior) 

Fixes #66

### CODE REVIEW

**Comment by nickva** [comment]
Test results:

```
        triq_tests:263: intrange_shrinking_test_...[0.002 s] ok
        triq_tests:264: intrange_shrinking_test_...[0.003 s] ok
        triq_tests:265: intrange_shrinking_test_...[0.002 s] ok

Ran 100 tests
[0.150 s] ok
=======================================================
  All 33 tests passed.
```

**Comment by tuncer** [comment]
+1 if we're certain that we should allow it to go below 0.

File: src/triq_dom.erl [hunk 1]
```diff
@@ -604,9 +604,9 @@ int(Max) ->
 int(Min, Max) ->
     Diff = Max - Min,
     #?DOM{kind=int,
-          shrink=fun(Dom,Val) when Val>0 -> {Dom,Val-1};
-                    (Dom,Val) when Val<0 -> {Dom,Val+1};
-                    (Dom,0) -> {Dom,0}
+          shrink=fun(Dom,Val) when Val>Min -> {Dom,Val-1};
+                    (Dom,Val) when Val<Min -> {Dom,Val+1};
```
Comment (KrzysiekJ) [inline src/triq_dom.erl]: Will this clause be ever met?

File: src/triq_dom.erl [hunk 2]
```diff
@@ -604,9 +604,9 @@ int(Max) ->
...
+                    (Dom,Val) when Val=:=Min -> {Dom,Val}
```
Comment (KrzysiekJ) [inline src/triq_dom.erl]: If we remove the clause above, then we should be able to rewrite the shrinking function as:

```erlang
fun(Dom,Min) -> {Dom,Min};
   (Dom,Val) -> {Dom,Val-1}
```

**Comment by KrzysiekJ** [comment]
I wonder whether we should keep separate `int/2` and `choose/2` generators, knowing that they do roughly the same (shrinking functions differ). QuickCheck has no `int/2` and PropEr has `integer/2` with documentation of `choose/2` saying: “Equivalent to `integer(Low, High)`.”.

**Comment by nickva** [comment]
Thanks for taking a look!

Looking at the other property testing tool's API, noticed in case of integer/2 the behavior is to shrink towards 0 if Min =< 0 =< Max and towards the value with smallest absolute distance to 0.

So [-2, 2] would shrink to 0, [-4,-2] to -2 and [2, 4] to 2. Then it makes sense to have a separate one from choose?

I'll update the PR to follow that logic I think.

File: src/triq_dom.erl (hunk 2 above)
Comment (nickva) [reply src/triq_dom.erl]: Agreed. Good point. I'll remove it and update it as suggested.

Comment (nickva) [reply src/triq_dom.erl]: ~Agreed. Good point. I'll remove it and update it as suggested.~

Looking at it a bit more, this clause is needed when the value we are shrinking towards is negative.

**Comment by KrzysiekJ** [comment]
But in QuickCheck there is no `integer/2` and in PropEr it is explicitly said that `integer/2` is equivalent to `choose/2`. So neither of these has two bounded integer generators that differ in shrinking target. Also in QuickCheck `choose/2` shrinks towards 0 (I guess that in PropEr too), so we should probably mimic that (currently `choose/2` in Triq shrinks towards minimum value).

**Comment by nickva** [comment]
Ah interesting. I took a better look. Here is what I found.

**QuickCheck**

QuickCheck has both a `choose/2` and `shrink_int/3` functions

http://quviq.com/documentation/eqc/eqc_gen.html#choose-2 it shrinks to the smallest absolute value. So `[-2, 4]` would shrink to `-2` it seems (as opposed to `0`)? `[-4,-2]` would also shrink to `-2`. I don't have access to QuickCheck to try it out though.

http://quviq.com/documentation/eqc/eqc_gen.html#shrink_int-3 : 

> Generates the value X, and shrinks it within the range M to N. The shrinking strategy used is the same one used by built - in QuickCheck generators.

I think it means it tries to do whatever integer does (shrink toward 0) but is constrained by the range limits. So `[-2, 4]` would shrink to `0` instead of `-2`. `[-4, -2]` to `-2`, and `[2, 4]` to `2`. 

**Proper**

Proper http://proper.softlab.ntua.gr/doc/ has both `choose/2` and `integer/2` but both are aliased to `integer/2`. 

> [...] Instances shrink towards 0 if Low =< 0 =< High, or towards the bound with the smallest absolute value otherwise.

I tried it and it seems to confirm it:

```
> proper_gen:sampleshrink(proper_types:integer(-4, 2)).
...
0
```

```
> proper_gen:sampleshrink(proper_types:integer(-4, -2)).
-2
```

**Triq**

Triq's `choose/2` currently seems to shrink to smallest value (not absolute value!). 

```
> triq_dom:sampleshrink(triq_dom:choose(-4,2)).
...
[-4]
```

```
triq_dom:sampleshrink(triq_dom:choose(-4,-2)).
[-4]
```

The `int/2` shrinking in this PR behaves like `shrink_int/3` in QuickCheck and `integer/2` in Proper. 

```
> triq_dom:sampleshrink(triq_dom:int(-4,-2)).
[-2]
```

```
4> triq_dom:sampleshrink(triq_dom:int(-4,2)).
[0]
```

Another interesting aside, noticed in Triq, the shrinking algorithm for `int/*` and `choose/2` is different in how fast it shrinks. `choose/2` bisects towards the minimum value, so it does exponential shrinking which is good for large intervals, but might skip large ranges in between. `int/*` does `+/-1` so it tries every integer, which might be good for small ranges (say choosing the number of branches in a tree). I guess I am 

So to summarize, maybe it is worth having this new `int/2` implementation to match Proper's integer/2 and QuickCheck's `shrink_int/3` ? Then also have a separate fix to make `choose/2` behave like QuickCheck's (absolute value vs minimum value)?

**Comment by KrzysiekJ** [comment]
`choose/2` in QuickCheck shrinks towards the smallest absolute value, which means that it shrinks to 0 in case of [-2, 4], not to -2.

PropEr’s documentation for `choose/2` may be misleading, because it always shrinks towards the smallest absolute value (which doesn’t contradict what is written).

**Comment by nickva** [comment]
@KrzysiekJ 

You're right. I've been interpreting "shrink to smallest absolute value" as "picks one of abs(Min) or abs(Max)" and not abs(any number in the interval between Min and Max). 

Got a hold of eqc mini and tried it as well:

```
> eqc_gen:sampleshrink(eqc_gen:choose(-4, 2)).
-3
--> [0,2,-1,-2]
```

So what do you and @tuncer think, then?

 * Merge this PR to fix `int/2` and have it behave like `choose/2` from QuickCheck and Proper. Then have a separate PR to fix `choose/2` to match (but maybe keep it with its separate more aggressive shrinking strategy?)

 * Combine choose/2 and int/2 (alias one to another). But then what to do about the shrinking strategy? Pick one or the other? Or maybe even switch based on the size of the interval -- if less than 100 switch to +/-1 otherwise do the /2 

 * Merge this PR to fix `int/2` to behave like integer/2 in Proper, but keep `choose/2` doing what it does now as I could see always shrinking to the lower value as useful.

**Comment by tuncer** [comment]
Regarding how aggressive shrinking is for `int/2`/`choose/2`, would it make sense to take the size of the range into account when considering the steps?

**Comment by nickva** [comment]
I was looking at `int()` and it has the same +/-1 strategy. Maybe the idea is that all `int/*` generators are for small integers and those are always shrunk with +/-1 steps. And `choose/2` is more suitable for large intervals so users would use them accordingly?

But if we wanted to unify them, (and just have choose or alias int to choose) then maybe have a threshold of say 20 and if range is larger than that use halving, if less use simple increments/decrements. This decision can be done when generator is initialized (by looking at `Diff=Max-Min`) or during each shrinking step by looking at the distance to the shrink target from current value.

**Comment by KrzysiekJ** [comment]
How about performing shrinking step randomly: half of the times by halving and half of the times by decrementing, like this:

```erlang
case triq_rnd:uniform(2) of
    1 ->
        %% Shrink by halving.
    2 ->
        %% Shrink by decrementing.
end
```
?

**Comment by tuncer** [comment]
Not sure that's a good idea, but wouldn't this, without reusing seeds, only be half of the time for the set of properties during one triq run?

**Comment by tuncer** [comment]
It sounds reasonable to shrink differently in `choose` and `int`, and I also think that taking the range into account is a good idea. If you have 1-1.000.000, you don't want to shrink by steps of 1.

Also, isn't shrinking configurable per property, so that some cases can be told to perform more steps?

**Comment by KrzysiekJ** [comment]
Not sure if I understand the question, but [there are many shrinking attempts](https://github.com/triqng/triq/blob/99d795e7d53d6d9bd185680bcc0a570f816a0eef/src/triq_dom.erl#L40-L41), so if we pick a mixed strategy, then likely some of the attempts will perform halving and some of them will perform decrementing. As a third option, we can subtract a random number.

In regard to shrinking target, it is easy to write a custom generator that shrinks towards the lowest value:

```erlang
choose_shrinking_to_lowest(Low, High) ->
    ?LET(
       Offset,
       choose(0, High - Low),
       Low + Offset).
```

I’m not saying however that we definitely should not have such generator as a built-in. In such case, it should probably be `int/2`, as `choose/2` in other libraries already shrinks to the lowest absolute value. But are there many use cases for such generator?

**Comment by hlieberman** [comment]
For what it's worth, we already support `largeint/0` in terms of exponential shrinking.  I'm not convinced we should worry too much about what happens if someone passes a large number into int/0, but I'm also not sure it's really /necessary/ for us to increment/decrement by one.

**Comment by nickva** [comment]
I like the random choice between decrementing and halving. Perhaps we can keep the behavior compatible (decrements when int value is small) but also handle the large values if around "interesting" values we do decrementing with a higher probability otherwise do halving.

"Interesting" values are the two boundaries plus the shrink target. Then for example, for `int(1, 10)` it might do mostly decrementing for any value. But for `int(1,100)`, if value is 80 or 54 it would more likely halve it. Then when it gets close to 1 it would start to decrement by 1.

I'll give that a try and update the PR and we can see how it looks.

@KrzysiekJ Good point on not needing a `choose` version which shrinks to lower limit. Example demonstrated well that it can be built easily.

**Comment by nickva** [comment]
I gave it a try and update the PR. `int/2` and `choose/2` now both do the same thing and know how to handle large and small intervals.

It takes into consideration @KrzysiekJ's idea of changing strategy of show shrinking is done and mixing the two strategies, but with an improvement that it is trying to consider values close to "interesting" boundaries using decrements. This also preserves the old int behavior for small ints where it decrements by 1:

```
triq_dom:sampleshrink(triq_dom:choose(1,5)).
5
[4]
[3]
[2]
[1]
```

```
triq_dom:sampleshrink(triq_dom:int(1,1 bsl 32)).
3332009066
[1666004533]
[833002267]
[416501134]
[208250567]
[104125284]
[52062642]
[26031321]
[13015661]
[6507831]
[3253916]
[1626958]
...
[6,4]
[5,3]
[4]
[3]
[2]
[1]
```

For a smooth transition between probability of using decrements vs increments used an exponential function. Not sure if that's overkill or not. It could have been a table probably as well.

**Comment by tuncer** [comment]
LGTM and tests and dialyzer pass. I don't think it modifies the API but, to be sure, let me ask: does the new record field (`shrinkto`) affect the API? If not, this looks ready to merge.

**Comment by nickva** [comment]
Modifying the choose record doesn't modify the API. It seems to be private to the triq_dom module.

**Comment by nickva** [comment]
@KrzysiekJ what do you think, I'd like your opinion as well if you get a chance to take a look

**Comment by nickva** [comment]
@hlieberman 

I was mostly convinced by Tuncer's argument that a user could pass in a large limit and then it would break unexpectedly. They should read the docs but it's easier perhaps if the code does something reasonable to start with.

Then with Kryzstof's suggestion of switching the strategy it didn't seem to difficult to implement.

**Comment by hlieberman** [comment]
I think it's fine; we may need to tweak the exponential at some point down the line, but I think it's a decent approach.

File: src/triq_dom.erl [hunk 3]
```diff
@@ -1194,9 +1190,30 @@ choose_pick(#?DOM{kind=#choose{min=M,max=N}}=Dom, _) ->
...
 
-choose_shrink(#?DOM{kind=#choose{min=M}}=Dom, Value) ->
-    Mid = (Value - M) div 2,
-    {Dom, M + Mid}.
+choose_shrink(#?DOM{kind=#choose{shrinkto=Value}}=Dom, Value) ->
+    {Dom, Value};
+choose_shrink(#?DOM{kind=#choose{min=M,max=N,shrinkto=ShrinkTo}}=Dom, Value) ->
+    ClosestBoundary = lists:min([M, N, ShrinkTo]),
+    DistToBoundary = abs(Value - ClosestBoundary),
+    DecrementProb = 10 * math:exp(-0.5 * DistToBoundary),
```
Comment (KrzysiekJ) [inline src/triq_dom.erl]: So if `DistToBoundary` is `20`, then `DecrementProb` will be about 0.045% . Should it not be higher? Looks like there is practically little chance to do shrinking by decrementing:

```
9> triq_dom:sampleshrink(triq_dom:choose(-1 bsl 40, 1 bsl 40)).
-581278805425
[-290639402712]
[-145319701356]
[-72659850678]
[-36329925339]
[-18164962669]
[-9082481334]
[-4541240667]
[-2270620333]
[-1135310166]
[-567655083]
[-283827541]
[-141913770]
[-70956885]
[-35478442]
[-17739221]
[-8869610]
[-4434805]
[-2217402]
[-1108701]
[-554350]
[-277175]
[-138587]
[-69293]
[-34646]
[-17323]
[-8661]
[-4330]
[-2165]
[-1082]
[-541]
[-270]
[-135]
[-67]
[-33]
[-16]
[-8]
[-4]
[-2]
[-1]
[0]
```

File: src/triq_dom.erl [hunk 4]
```diff
@@ -1194,9 +1190,30 @@ choose_pick(#?DOM{kind=#choose{min=M,max=N}}=Dom, _) ->
     Value = triq_rnd:uniform(N-M+1) - 1 + M,
     {Dom,Value}.
 
-choose_shrink(#?DOM{kind=#choose{min=M}}=Dom, Value) ->
-    Mid = (Value - M) div 2,
-    {Dom, M + Mid}.
+choose_shrink(#?DOM{kind=#choose{shrinkto=Value}}=Dom, Value) ->
+    {Dom, Value};
+choose_shrink(#?DOM{kind=#choose{min=M,max=N,shrinkto=ShrinkTo}}=Dom, Value) ->
+    ClosestBoundary = lists:min([M, N, ShrinkTo]),
```
Comment (KrzysiekJ) [inline src/triq_dom.erl]: From the definition it looks like it is actually `LowestBoundary` hence `M`. Why don’t we measure distance just in relation to `ShrinkTo`?

Comment (nickva) [reply src/triq_dom.erl]: Oops. Each member in the list should have been `abs(Value - ...)`

But I'll switched it to measure the distance just to ShrinkTo then.

**Comment by tuncer** [comment]
I plan on doing the gitlab migration (#31) today or tomorrow, so it'd be great if we could merge this soon. If there's room for further improvement/adjustments, we can work it out in a 2nd patch. With GitLab-CI in place, it'll be easier.

File: src/triq_dom.erl (hunk 3 above)
Comment (nickva) [reply src/triq_dom.erl]: How about these coefficients. I'll tweak the coefficients to always decrement for about 1-5 then by 20 or so make it 10%. 
 
For example:
```
[{X, 4*math:exp(-0.2*X)} || X <- lists:seq(1, 20)].
[{1,3.2749230123119273},
 {2,2.6812801841425573},
 {3,2.1952465443761056},
 {4,1.7973158564688863},
 {5,1.4715177646857693},
 {6,1.204776847648808},
 {7,0.9863878557664257},
 {8,0.8075860719786215},
 {9,0.6611955528863461},
 {10,0.5413411329464508},
 {11,0.4432126334493355},
 {12,0.3628718131576499},
 {13,0.2970943128573355},
 {14,0.2432402505008718},
 {15,0.19914827347145578},
 {16,0.16304881591346485},
 {17,0.13349307984130426},
 {18,0.10929488978917024},
 {19,0.08948308742466236},
 {20,0.07326255555493671}]
```

**Comment by nickva** [comment]
@KrzysiekJ Thanks for taking a look.

I updated it to only consider ShrinkTo target when deciding to pick decrement vs halving.

Tweaked the coefficients to make it always do decrement up to about 7. By 20 the chance of decrement is about 30%

```
[{X, 2*math:exp(-0.1*X)} || X <- lists:seq(1, 20)].
[{1,1.809674836071919},
 {2,1.6374615061559636},
 {3,1.4816364413634358},
 {4,1.3406400920712787},
 {5,1.2130613194252668},
 {6,1.0976232721880528},
 {7,0.9931706075828189},
 {8,0.8986579282344431},
 {9,0.8131393194811982},
 {10,0.7357588823428847},
 {11,0.6657421673961591},
 {12,0.602388423824404},
 {13,0.5450635860680252},
 {14,0.49319392788321287},
 {15,0.44626032029685964},
 {16,0.40379303598931077},
 {17,0.3653670481054692},
 {18,0.33059777644317306},
 {19,0.29913723844527007},
 {20,0.2706705664732254}]
```

This should get closer to what an int() does where it picks small values usually around 10-20 then decrements them by 1. 

```
 triq_dom:sampleshrink(triq_dom:choose(-1 bsl 40,1 bsl 40)).
triq_dom:sampleshrink(triq_dom:choose(-1 bsl 40,1 bsl 40)).
-294315490751
[-147157745375]
[-73578872687]
[-36789436343]
[-18394718171]
[-9197359085]
[-4598679542]
[-2299339771]
[-1149669885]
[-574834942]
[-287417471]
[-143708735]
[-71854367]
[-35927183]
[-17963591]
[-8981795]
[-4490897]
[-2245448]
[-1122724]
[-561362]
[-280681]
[-140340]
[-70170]
[-35085]
[-17542]
[-8771]
[-4385]
[-2192]
[-1096]
[-548]
[-274]
[-137]
[-68]
[-34]
[-17]
[-16,-8]
[-8,-15]
[-7]
[-6]
[-5]
[-4]
[-3]
[-2]
[-1]
[0]
ok
```

**Comment by nickva** [comment]
@tuncer I think I addressed the last issues @KrzysiekJ  pointed out. Tests still pass. Can merge if it looks good.

**Comment by tuncer** [comment]
Since the merge, `non_empty_list_shrink_test` times out. From [51347573](https://gitlab.com/triq/triq/-/jobs/51347573):

```
file "triq.app"
  application 'triq'
    module 'triq'
      module 'triq_tests'
        triq_tests: boolean_test...[0.072 s] ok
        triq_tests: delete_test...[0.491 s] ok
        triq_tests: suchthat_sample_size_test...[0.902 s] ok
        triq_tests: tuple_failure_test...[0.411 s] ok
        triq_tests: list_shrink_test...[0.946 s] ok
        triq_tests: non_empty_list_shrink_test...*timed out*
undefined
```

**Comment by nickva** [comment]
@tuncer Taking a look. 

It passes locally for me in about 2 seconds. Wonder if it is an Erlang version deal. I am using 20.0

Saw 17 through 20 in the CI dot file but wondering about this particular job

**Comment by nickva** [comment]
I tried with 17, 19 and 20 locally multiple times. I think it takes about 2 seconds on my laptop and the CI environment has spiky performance (CPU throttling) I could see timeouts being an issue. Default timeout in EUnit is 5 seconds.

Looked at the test code as well and it seems `non_empty_shrink_test` tests list() and int() so it should haven't been affected by this change.

We could reduce the number of tries from 25 to 10. Or use a timeout setup option:
```
non_empty_list_shrink_test_() ->
   {timeout, 20, ?_assert(...)}.
```

**Comment by tuncer** [comment]
You're right. It's a shared resource and it's very likely that, even though the latest commit isn't related, it might have changed the odds a little bit. Either option works for me, but let's test reducing tries first.

**Comment by nickva** [comment]
@tuncer PR to reduce number of tries from 25 to 15: https://github.com/triqng/triq/pull/74

**Comment by tuncer** [comment]
Merged.
